src/lib/tls/asio/asio_error.h

```diff
@@ -0,0 +1,176 @@
+/*
+* TLS Stream Errors
+* (C) 2018-2019 Jack Lloyd
+*     2018-2019 Hannes Rantzsch, Tim Oesterreich, Rene Meusel
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_ASIO_ERROR_H_
+#define BOTAN_ASIO_ERROR_H_
+
+#include <botan/build.h>
+
+#if defined(BOTAN_HAS_TLS) && defined(BOTAN_HAS_BOOST_ASIO)
+
+#include <boost/version.hpp>
+#if BOOST_VERSION >= 106600
+
+#include <boost/system/system_error.hpp>
+
+#include <botan/tls_alert.h>
+
+namespace Botan {
+
+namespace TLS {
+
+enum class error
+   {
+   unexpected_message = 1,
+   invalid_argument,
+   unsupported_argument,
+   invalid_state,
+   key_not_set,
+   lookup_error,
+   internal_error,
+   invalid_key_length,
+   invalid_iv_length,
+   prng_unseeded,
+   policy_violation,
+   algorithm_not_found,
+   no_provider_found,
+   provider_not_found,
+   invalid_algorithm_name,
+   encoding_error,
+   decoding_error,
+   integrity_failure,
+   invalid_oid,
+   stream_io_error,
+   self_test_failure,
+   not_implemented,
+   unknown
+   };
+
+namespace detail {
+// TLS Alerts
+struct BotanAlertCategory : boost::system::error_category
```

Comment:
That's not easy, unfortunately. `error_code` creation relies on providing overloads for `make_error_code` and we cannot define both `std::error_code make_error_code(...)` and `boost::error:code make_error_code(...)` (from the same enum class type). Hence, if we changed all this code to work on `std::error_code`, we'd no longer be able to automatically create `boost::error_code` from our enums.

We need to create `boost::error_code` because asio relies on them in its read and write methods. Of course we could have the internal machinery to automatically create `std::error_code` and then cast to `boost::error_code` wherever necessary. Just makes the code duplication discussed above larger, but maybe we can come up with a solution for this as well.

Reply:
OK thanks for explanation, in that case let's just stick with the Boost versions for now.